keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,172 @@
+owning-sig: sig-instrumentation
+reviewers:
+  - "@Random-Liu"
+  - "@bogdandrutu"
+approvers:
+  - "@brancz"
+creation-date: 2018-12-04
+last-updated: 2018-12-07
+status: provisional
+---
+
+# Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles
+
+## Table of Contents
+
+- [Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles](#leveraging-distributed-tracing-to-understand-kubernetes-object-lifecycles)
+    - [Table of Contents](#table-of-contents)
+    - [Summary](#summary)
+    - [Motivation](#motivation)
+        - [Goals](#goals)
+        - [Non-Goals](#non-goals)
+    - [Proposal](#proposal)
+        - [Architecture](#architecture)
+            - [Trace lifecycle](#trace-lifecycle)
+            - [Root spans](#root-spans)          
+            - [Context propagation](#context-propagation)          
+        - [Out-of-tree changes](#out-of-tree-changes)
+            - [Mutating admission webhook](#mutating-admission-webhook)
+            - [Tracing best-practices documentation](#tracing-best-practices-documentation)
+        - [In-tree changes](#in-tree-changes)
+            - [Trace utility package](#trace-utility-package)
+            - [Basic Kubernetes component trace instrumentation](#basic-kubernetes-component-trace-instrumentation)
+    - [Graduation Requirements](#graduation-requirements)
+    - [Implementation History](#implementation-history)
+
+## Summary
+
+This Kubernetes Enchancement Proposal (KEP) introduces a model for adding distributed tracing to Kubernetes object lifecycles. The inclusion of this trace instrumentation will mark a significant step in making Kubernetes processes more observable, understandable, and debuggable.
+
+
+## Motivation
+
+Debugging latency issues in Kubernetes is an involved process. There are existing tools which can be used to isolate these 
+issues in Kubernetes, but these methods fall short for various reasons. For instance:
+
+* **Logs** are fragmented, and finding out which process was the bottleneck involves digging through troves of unstructured text. In addition, logs do not offer higher-level insight into overall system behavior without an extensive background on the process of interest. 
+* **Events** in Kubernetes are only kept for an hour by default, and don't integrate with visualization of analysis tools. To gain trace-like insights would require a large investment in custom tooling
+* **Latency metrics** are gathered in some places, but these don't provide understanding into _why_ a given process was slow
+
+Distributed tracing, on the other hand, provides a single window into latency information from across many components and plugins. Trace data is structured, and there are numerous established backends for visualizing and querying over it. This KEP would make it possible to, for instance, retrieve and visualize all pod startups that took more than 30 seconds, involved an `nginx` container, and which mounted more than two volumes.
```

Comment:
> and which mounted more than two volumes.

any particular reason why this is important? I can't tell if you're putting emphasis on the querying capability of tracing solutions or the detail of spans this is intended to create. The latter would be important to specify more clearly, if it's just about querying capabilities then the first sentence of this paragraph is sufficient I think.

Reply:
:+1: revised